src/coreclr/jit/importercalls.cpp

```diff
@@ -7558,15 +7558,10 @@ CORINFO_CLASS_HANDLE Compiler::impGetSpecialIntrinsicExactReturnType(GenTreeCall
 
             // Lookup can incorrect when we have __Canon as it won't appear
             // to implement any interface types.
-            //
-            // And if we do not have a final type, devirt & inlining is
-            // unlikely to result in much simplification.
-            //
-            // We can use CORINFO_FLG_FINAL to screen out both of these cases.
-            const DWORD typeAttribs = info.compCompHnd->getClassAttribs(typeHnd);
-            bool        isFinalType = ((typeAttribs & CORINFO_FLG_FINAL) != 0);
+            bool isCanon = ((info.compCompHnd->getClassAttribs(typeHnd) &
+                             (CORINFO_FLG_SHAREDINST | CORINFO_FLG_FINAL)) == CORINFO_FLG_SHAREDINST);
```

Comment:
This is getting too complicated. Let's first agree how this should work in plain English. I think it should be:

1. If the `call` target is shared generic code, get more precise type from instantiation argument if possible. If it is not possible that's ok, we will work with what we got.

2. Call JIT/EE interface to ask for the exact comparer type. If the VM cannot provide exact comparer type (e.g. the type is __Canon), it should return NULL.

Q: Can shared generic types like `ObjectEqualityComparer<MyStruct<__Canon>>` be treated as exact by the JIT today? In other words, should the JIT/EE interface call bail on shared generic types or should it try to return shared generic type where possible?

Reply:
> Can shared generic types like `ObjectEqualityComparer<MyStruct<__Canon>>` be treated as exact by the JIT today?

The JIT already treats those as exact. I think that moving the detection to VM like you proposed is a good idea

Reply:
Sounds good. Could you please implement this plan?

- Both `getClassAttribs` calls in this method should be deleted. We can just look for WellKnownArg::InstParam and use it if it is available. It is cheaper than `getClassAttribs` call.
- `getDefaultEqualityComparerClass` and `getDefaultComparerClass` implementation should be fixed up to do the above.